File: December 2022/27 December 2022/C++/Container With Most Water.cpp
//{ Driver Code Starts
//Initial template for C++

#include<bits/stdc++.h>
using namespace std;
// int maxArea(int A[], int len);

// } Driver Code Ends

long long maxArea(long long A[], int len)
{
	int l = 0;
	int r = len -1;
	long long area = 0;

	while (l < r)
	{
		area = max(area, min(A[l], A[r]) * (r - l));

		if (A[l] < A[r])
			l += 1;
		else
			r -= 1;
	}
	return area;
}

//{ Driver Code Starts.

// Driver code
int main()
{
	int t;
	cin>>t;
	while(t--)
    {
        int n;
        cin>>n;
        long long arr[n];
        for(int i=0;i<n;i++)
            cin>>arr[i];
        cout<<maxArea(arr,n)<<endl;
    }
return 0;
}

// } Driver Code Ends

File: December 2022/27 December 2022/JavaScript/Container With Most Water.js
//{ Driver Code Starts
//Initial Template for javascript

'use strict';

process.stdin.resume();
process.stdin.setEncoding('utf-8');

let inputString = '';
let currentLine = 0;

process.stdin.on('data', inputStdin => {
    inputString += inputStdin;
});

process.stdin.on('end', _ => {
    inputString = inputString.trim().split('\n').map(string => {
        return string.trim();
    });
    
    main();    
});

function readLine() {
    return inputString[currentLine++];
}

function printArray(res, size) {
    for(let i=0;i<size;i++){
        console.log(res[i]);
    }
}

function main() {
    let t = parseInt(readLine());
    let i = 0;
    for(;i<t;i++)
    {
        let n = parseInt(readLine());
        let arr = new Array(n);
        let input_ar1 = readLine().split(' ').map(x=>parseInt(x));
        for(let i=0;i<n;i++){
            arr[i] = input_ar1[i];
        }
        let obj = new Solution();
        let res = obj.maxArea(arr, n);
        console.log(res);
    }
}

// } Driver Code Ends

class Solution {
    maxArea(A, len){
        let l = 0;
    	let r = len -1;
    	let area = 0;
    
    	while (l < r)
    	{
    		area = Math.max(area, Math.min(A[l], A[r]) * (r - l));

    		if (A[l] < A[r])
    			l += 1;
    		else
    			r -= 1;
    	}
    	return area;
    }
}